[1단계 - 로또(자동)] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
### 안녕하세요 닉!! 디우라고 합니다. ㅎㅎ

---

한 가지 질문드리고 싶은 것이 있습니다!
`WinningNumbers` 라고 하는 당첨 번호 List 를 관리하는 일급 컬렉션을 두고, 정적 팩토리 메소드를 적용하여 아래와 같이 만들어 보았습니다.

```java
private WinningNumbers(List<Integer> normalWinningNumbers, Integer bonusWinningNumber) {
        validateNormalWinningNumbers(normalWinningNumbers);
        validateDuplication(normalWinningNumbers, bonusWinningNumber);

        this.winningNumbers = getWinningNumbers(normalWinningNumbers, bonusWinningNumber);
    }

...

public static WinningNumbers create(List<String> normalWinningValues, String bonusWinningValue) {
        List<Integer> normalWinningNumbers = translateIntegerList(normalWinningValues);
        Integer bonusWinningNumber = translateInteger(bonusWinningValue);

        return new WinningNumbers(normalWinningNumbers, bonusWinningNumber);
    }
```

이 때 생성자에서 검증에 대한 로직을 두고, 정적 팩토리 메소드에서 변환을 수행하다 보니 도메인 자체가 너무 무거워진거 아닌가 하는 생각이 들었습니다. 책임의 측면에서 보면 검증에 대한 로직은 해당 도메인에 두는 것이 맞다고 생각되는데 변환과 같은 부분은 별도로 둘 수 있을 것 같다는 생각이 들었습니다. 그렇다면 해당 로직은 입력을 받는 View로 이동시켜야할까요..?? 어디에 두면 좋을지 모르겠습니다..혹은 지금 그대로 괜찮은지 궁금합니다...!!

---

이전 미션에 비해서 많이 복잡해져서 페어와 함께 도메인을 단계별로 정리해가며 구현해보았습니다.
그래도 아직 코드에서 개선할 부분이 많다고 생각됩니다...따끔한 피드백 부탁드립니다..!!

hyeonic committed:
docs(README): 페어 프로그래밍 규칙 및 기능 구현 목록 추가

hyeonic committed:
feat: Amount 티켓 발급 기능

투입 금액 기반으로 티켓 발금 기능 구현

hyeonic committed:
feat: Amount 구입 금액이 부족한 경우 예외 처리 추가

hyeonic committed:
refactor: Amount 누락된 상수 추가

hyeonic committed:
feat: Amount 거스름돈 있는 경우 추가

거스름돈이 생기는 경우 가능한 티켓 갯수만큼 반환

hyeonic committed:
test: Amount 음수 입력 검증

구입금액이 음수인 경우 예외를 던지는지 검증

hyeonic committed:
feat: LottoTicket 생성 기능 추가

LottoTicket 생성 시점에 로또 번호 생성을 위한 전략 활용

hyeonic committed:
feat: LottoNumberGenerator 로또 번호 생성 전략 추가

자동으로 로또 번호 생성을 위한 AutoLottoNumberGenerator 추가 및 검증

hyeonic committed:
feat: LottoTickets 일급컬렉션 추가

LottoTicket 리스트를 관리하는 일급컬렉션을 추가

hyeonic committed:
feat: LottoTicket 불변을 위한 기능 추가

Collections.unmodifiableList 활용

hyeonic committed:
feat: Amount 정적 팩토리 메소드 적용

Amount 생성자를 private 으로 막고 정적 팩토리 메소드를 적용하여 create() 메소드를 통해 생성하도록 구현

hyeonic committed:
feat: LottoMachine 구현

price 를 받아 로또 티켓을 자동으로 발급하도록 구현

hyeonic committed:
feat: LottoTickets 티켓 갯수 반환 기능 추가

hyeonic committed:
feat: LottoController 구매 과정 및 로또 정보 조회 추가

hyeonic committed:
feat: WinningNumber 당첨 번호 기능 추가

당첨 번호 한 자리의 책임을 가진 WinningNumber 추가

hyeonic committed:
feat: WinningNumbers 당첨 번호 생성 확인

예외 없이 제대로 WinningNumbers 를 생성하는지 확인

hyeonic committed:
feat: 당첨번호 컨트롤러 구현

당첨번호 및 보너스 볼을 처리하는 컨트롤러 구현

hyeonic committed:
refactor: LottoController 메서드 기능 분리

hyeonic committed:
feat: WinningNumber 당첨 번호 확인을 위한 기능 추가

당첨 번호를 포함했지는 유무를 판단한다.

hyeonic committed:
feat: 당첨 정보 열거형 생성

당첨 정보에 해당하는 열거형(Rank)를 생성하고 제대로 반환하는지 테스트

hyeonic committed:
test: 당첨 정보의 검증을 위한 테스트 추가

당첨 정보에 해당하는 열거형(Rank)를 생성하고 제대로 반환하는지 테스트

hyeonic committed:
feat: WinningNumbers 당첨 정보 반환 기능 추가

LottoTicket을 기반으로 비교하여 당첨 정보를 반환한다.

hyeonic committed:
feat: 수익률 계산 구현

LottoResult에서 투입 금액을 받아 수익률 계산을 하도록 구현

hyeonic committed:
feat: 로또 결과 구하는 메소드 추가

WinningNumbers 를 받아 로또 결과를 구하는 메소드를 구현

hyeonic committed:
feat: LottoController 당첨 통계 구현

hyeonic committed:
refactor: Amount -> Money 네이밍 변경

hyeonic committed:
refactor: 로또 번호가 당첨 번호와 같은지 묻는 주체 변경

인자로 넘겨져오는 LottoTicket에 메시지를 보내서 같은 번호인지를 묻도록 변경

hyeonic committed:
feat: 입출력 처리를 위한 InputView, OutputView 추가

LottoController 생성자 주입을 통해 view 활용

hyeonic committed:
refactor: 로또 티켓 정보 출력 view로 책임 이동

hyeonic committed:
refactor: 로또 당첨 번호 view 로 책임 이동

hyeonic committed:
refactor: 로또 결과 출력 안내문 view 로 분리

hyeonic committed:
refactor: 통계 출력 부분 view로 책임 이전

tco0427 committed:
refactor: Money에 정적 팩토리 메소드 적용

정적 팩토리 메소드를 적용함으로 컨트롤러쪽의 createMoney() 메소드 부분도 함께 수정하였습니다.

tco0427 committed:
feat: LottoTicketDto 생성

view 에서 domain 객체를 직접 사용하는 것 보다는 DTO를 만들어 사용하는게 적절하다고 생각하여 LottoTicketDto 를 만들었습니다.

tco0427 committed:
feat: LottoTicketsDto 생성

view 에서 domain 객체를 직접 사용하는 것 보다는 DTO 를 만들어 사용하는게 적절하다고 생각하여 LottoTicketsDto 를 만들었습니다.

tco0427 committed:
refactor: view 에서 domain 의존 -> dto 의존 하도록 수정

DTO 를 넘겨 view 가 domain 이 아닌 dto 에 의존하도록 수정하였습니다.

tco0427 committed:
refactor: LottoController 예외 범위 수정

NumberFormatException 을 catch 하도록 수정하였고, 메시지를 추가해주었습니다.

tco0427 committed:
feat: WinningNumbers 생성시 검증할 부분 추가

* 중복 검증
* 1~45의 숫자인지 검증
* 일반 당첨 번호가 6개인지 검증
* 정수인지를 검증

tco0427 committed:
style: 불필요한 import문 제거

불필요한 import 구문을 제거하고, 메소드 순서를 조율하였습니다.

tco0427 committed:
refactor: 정적 팩토리 메소드를 적용한 domain 의 생성자 private

정적 팩토리 메소드를 적용한 도메인에 대해서 private 생성자로 접근제어자 수정

tco0427 committed:
fix: 검증 과정에서 보너스 볼 2개 들어가는 문제 해결

중복 검증 과정에서 보너스 볼이 2개 들어가는 문제가 발생하여 이를 해결하였습니다.

tco0427 committed:
fix: @SuppressWarning("NonAsciiCharacters") 추가

tco0427 committed:
feat: 1~45 사이의 숫자인지 검증하는 코드를 WinningNumber로 이동

각각의 WinningNumber 에서 1~45 사이인지 검증하는 것이 보다 적절한 책임 분배라고 생각하여 리팩토링 하였습니다.
또 그에따른 테스트 코드를 작성하였습니다.

tco0427 committed:
docs(readme): WinningNumbers의 기능 추가 작성

중복되는 숫자가 있는지 검증하는 기능이 추가되었습니다.

tco0427 committed:
refactor: WinningNumbers 메소드 순서 조율

읽히는 순서대로 메소드 순서를 변경하였습니다.

tco0427 committed:
test: 기본 6개의 당첨 번호 중복 확인

기존 보너스볼의 번호 중복 이외에
기본 6개의 당첨 번호가 중복되는지를 테스트하였습니다.

tco0427 committed:
refactor: enum 디렉토리 분리

enumeration 패키지 하위에 두어 구분하도록 하였습니다.

tco0427 committed:
test: 정수가 아닌 당첨 번호 입력 테스트

정수가 아닌 당첨 번호 입력에 대해서 테스트해보았습니다.

tco0427 committed:
docs(readme): 프로젝트 상단으로 이동

README.md 파일을 프로젝트 상단으로 이동시켰습니다.

tco0427 committed:
docs(readme): 도메인 관계 그림 추가

도메인 간에 흐름을 이해할 수 있는 그림을 추가하였습니다.

tco0427 committed:
refactor: WinningNumbersTest 에서 공통되는 로직을 메소드로 추출

List<String> 생성에서 중복되는 로직을 별도의 메소드로 추출하였습니다.

wbluke commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+        Integer bonusWinningNumber = translateInteger(bonusWinningValue);
+
+        return new WinningNumbers(normalWinningNumbers, bonusWinningNumber);
+    }
```
> 한 가지 질문드리고 싶은 것이 있습니다!
WinningNumbers 라고 하는 당첨 번호 List 를 관리하는 일급 컬렉션을 두고, 정적 팩토리 메소드를 적용하여 아래와 같이 만들어 보았습니다.
> ```java
> private WinningNumbers(List<Integer> normalWinningNumbers, Integer bonusWinningNumber) {
>         validateNormalWinningNumbers(normalWinningNumbers);
>         validateDuplication(normalWinningNumbers, bonusWinningNumber);
> 
>         this.winningNumbers = getWinningNumbers(normalWinningNumbers, bonusWinningNumber);
>     }
> 
> ...
> 
> public static WinningNumbers create(List<String> normalWinningValues, String bonusWinningValue) {
>         List<Integer> normalWinningNumbers = translateIntegerList(normalWinningValues);
>         Integer bonusWinningNumber = translateInteger(bonusWinningValue);
> 
>         return new WinningNumbers(normalWinningNumbers, bonusWinningNumber);
>     }
> ```
> 이 때 생성자에서 검증에 대한 로직을 두고, 정적 팩토리 메소드에서 변환을 수행하다 보니 도메인 자체가 너무 무거워진거 아닌가 하는 생각이 들었습니다. 책임의 측면에서 보면 검증에 대한 로직은 해당 도메인에 두는 것이 맞다고 생각되는데 변환과 같은 부분은 별도로 둘 수 있을 것 같다는 생각이 들었습니다. 그렇다면 해당 로직은 입력을 받는 View로 이동시켜야할까요..?? 어디에 두면 좋을지 모르겠습니다..혹은 지금 그대로 괜찮은지 궁금합니다...!!

좋은 질문입니다 ㅎㅎ
여기서 고민해봐야 할 포인트는 다음과 같습니다.

**모든 validation이 다 도메인의 역할일까?**

저는 View라는 계층이 별도로 있다는 점에서, View에서 최소한의 사용자 입력 유효성은 보장해주어야 하지 않을까? 라고 생각해요.
우리가 로또 숫자 1개를 입력 받을 때 고려해야 할 validation은 다음과 같은 것들이 있죠.
1. 해당 입력이 숫자인가?
2. 해당 입력이 1과 45 사이의 숫자인가?

분명 1번 validation과 2번 validation은 성격이 다릅니다.
1번은 우리가 정한 도메인 요구사항과 관련 없이, 숫자를 원한다면 숫자를 받아야 하는 최소한의 값 정합성인 것이고,
2번은 우리가 정한 도메인 요구사항의 규칙에 맞는지 검사한 것이죠.

어떻게 수정하면 좋을지 한번 고민해보시고, 여기 뿐만 아니라 전체적으로 리팩토링 해보셔요 :)

tco0427 replied:
우선 `View에서 최소한의 사용자 입력 유효성을 보장` 해야한다는 말에 동의합니다. 따라서 String -> int 변환 및 검증에 대한 책임을 view 단으로 옮겼습니다. 이 때 view에서 변환된 int를 통해서 LottoNumber를 생성해 반환하도록 하였습니다.
`LottoNumber` VO 객체를 사용함으로써 1과 45 사이의 숫자인지를 검증하는 책임을 지도록 하였습니다.
따라서 뷰에서 LottoNumber를 생성해 반환해주는 과정에서 1번과 2번 모두 검증하게 됩니다!

* view : 숫자인지 검증
* LottoNumber : 1~45 사이의 숫자인지 검증

wbluke commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,41 @@
+    private Money(int price) {
+        validateCriterion(price);
+        this.price = price;
+    }
```
이름은 Money 인데, `로또를 구입하기 위한` 최소한의 금액을 보장해야 하는 Money 네요.
클라이언트가 해당 객체에 대한 정보 없이 500원이라는 돈을 생성하고자 하면 예상과 달리 예외가 발생할 거에요.
더 구체적인 네이밍을 고민해볼까요? :)

wbluke replied:
그리고 VO라면 기본적으로 eq & hc에 대한 재정의도 항상 같이 고민해주세요 :)
(VO : Value Object)

tco0427 replied:
동의합니다!! 단순 Money라고 하면 10원, 100원 생성도 가능해야할 것 같습니다. 따라서 "로또 구입을 위한 금액" 이라는 의미를 좀 더 느낄 수 있도록 LottoPurchaseMoney라는 이름으로 변경해주었습니다!
또한 해당 클래스는 '로또 구입 금액' 이라는 특정 값을 나타내는 객체인 VO 이므로 equals 와 hashCode 를 오버라이딩하여 재정의해주었습니다!

wbluke commented on src/main/java/lotto/domain/generator/AutoLottoNumberGenerator.java:
```diff
@@ -0,0 +1,29 @@
+
+public class AutoLottoNumberGenerator implements LottoNumberGenerator {
+
+    private static final List<Integer> LOTTO_TOTAL_NUMBERS = IntStream.rangeClosed(1, 45)
```
애플리케이션 전체적으로 로또 숫자가 Integer로만 관리되고 있네요.
LottoNumber라는 VO를 정의해보면 어떨까요?
그러면 1 ~ 45 사이의 validation도 자연스럽게 할 수 있고, 해당 상수 관리도 할 수 있을 것 같아요 :)

tco0427 replied:
좋은 의견이라고 생각합니다!!😃
말씀해주신 것 처럼 1~45 validation 책임도 LottoNumber로 위임할 수 있을 것 같습니다!! 수정해보았습니다!

wbluke commented on src/main/java/lotto/domain/generator/AutoLottoNumberGenerator.java:
```diff
@@ -0,0 +1,29 @@
+            .collect(toList());
+
+    private static final int LOTTO_NUMBERS_FROM_INDEX = 0;
+    private static final int LOTTO_NUMBERS_TO_INDEX = 6;
```
6이라는 숫자는 LottoTicket에서 관리되면 자연스럽지 않을까요? :)
(0이라는 인덱스는 굳이 상수화되지 않아도 될 것 같아요)

tco0427 replied:
해당 부분은 고민이 됩니다..ㅠㅠ
지금은 Lotto'Number'Generator라서 조금 더 의미가 불분명한 것 같은데, 만약 네이밍을 Lotto'Ticket'Generator 라고 한다면 6자리의 티켓 하나를 생성해준다는 의미에서 현재의 코드를 유지해줘도 되지 않나 라는 생각이 듭니다. 이것 또한 자연스럽지 않나..하는 생각도 듭니다..!!
(로또를 사러 가면 자동 생성기가 6자리의 숫자를 생성해 준다는 맥락에서....)

만약 말씀해주신 것 처럼 6이라는 매직넘버를 LottoTicket에서 관리하는 것도 자연스럽기는 하지만 그렇게 된다면 각각의 번호에 대한 생성 책임이 LottoTicket으로 넘어가게 될 것 같습니다...🥲

wbluke replied:
음 만약에 로또 게임이 7자리 숫자로 규칙이 변경된다면 결국 LottoTicket도 7개의 숫자, Generator도 7개의 숫자를 사용해야 하는 것이니 같은 상수 아닐까요?
그리고 `각각의 번호에 대한 생성 책임이 LottoTicket으로 넘어가게 될 것 같습니다...`라고 하셨는데, LottoTicket의 상수를 공개로 열고, Generator에서 사용하면 되지 않을까요? :)

wbluke commented on src/main/java/lotto/domain/LottoTicket.java:
```diff
@@ -0,0 +1,22 @@
+    private final List<Integer> lottoNumbers;
+
+    public LottoTicket(LottoNumberGenerator lottoNumberGenerator) {
+        this.lottoNumbers = lottoNumberGenerator.generate();
```
- 숫자가 6개인지
- 중복되는 숫자는 없는지

validation 할 수 있지 않을까요? :)

tco0427 replied:
맞습니다!! 수정해보도록하겠습니다~_~

wbluke commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,37 @@
+
+public class InputView {
+
+    private static final String SEPARATOR = ", ";
```
<img width="342" alt="image" src="https://user-images.githubusercontent.com/44018338/155825914-8b5fd92d-587e-414f-9071-d3e9ac4a0c43.png">

저는 잘못 입력하지 않았는데 위와 같은 결과가 나오고 있어요.
사용자 입력에 대한 유연함을 가지면 좋을 것 같아요 :)

tco0427 replied:
동의합니다!! 수정해보도록 하겠습니다..!!😄

wbluke commented on src/main/java/lotto/domain/enumeration/Rank.java:
```diff
@@ -0,0 +1,39 @@
+
+    private final int prizeMoney;
+    private final int count;
+    private final boolean isBonus;
```
해당 필드를 사용하고 있지 않네요? :)

tco0427 replied:
아래의 코멘트 주신 것처럼 outputView의 유연성을 위해 prizeMoney를 사용하도록 해보았습니다!!

wbluke replied:
엇 저는 isBonus를 사용하고 있지 않다고 말씀드린거에요

tco0427 replied:
오,,,어떤 말씀이신지 이해하였습니다..!! 수정해보겠습니다!

wbluke commented on src/main/java/lotto/domain/enumeration/Rank.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static Rank of(int count, boolean isBonus) {
+        if (isBonus && count == 5) {
```
5라는 숫자는 어떤 숫자일까요? :)

tco0427 replied:
5라는 숫자는 2등 혹은 3등일 때의 맞춘 번호 개수입니다..!!
이 때 25번 라인에서 2등을 검사하는 곳에서 사용하고 있으므로 그 의미를 명확히 할 수 있도록 `SECOND_PRICE_COUNT` 라는 상수로 치환해보았습니다!

wbluke replied:
음 이미 존재하고 있는 상수 아닐까요? ㅎㅎ
```java
if (isBonus && count == SECOND.count) {
```

wbluke commented on src/main/java/lotto/domain/enumeration/Rank.java:
```diff
@@ -0,0 +1,39 @@
+    THIRD(1_500_000, 5, false),
+    FORTH(50_000, 4, false),
+    FIFTH(5_000, 3, false),
+    NOTTING(0, 0, false);
```
`NOTHING` 이겠죠? 순간 제가 잘못 아는줄 알았네요 ㅋㅋ

tco0427 replied:
앗...!! 오타이네요ㅠㅠ 감사합니다!!😅

wbluke commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+        return winningNumbers.stream()
+                .filter(lottoTicket::isSame)
+                .collect(toList())
+                .size();
```
```suggestion
        return (int) winningNumbers.stream()
                .filter(lottoTicket::isSame)
                .count();
```

캐스팅 때문에 깔끔하진 않지만.. 불필요한 리스트는 생기지 않아요 :)

tco0427 replied:
리스트 생성에 대한 비용을 지불하지 않아도 돼서 보다 적절한 것 같습니다!! 꼼꼼한 피드백 감사드립니다~_~

wbluke commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private void validateNormalWinningNumbers(List<Integer> normalWinningNumbers) {
+        if (normalWinningNumbers.size() != 6) {
```
매직 넘버!

tco0427 replied:
앗..수정하도록 하겠습니다!!ㅎㅎ..

wbluke commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,28 @@
+
+        for (Rank rank : ranks.keySet()) {
+            totalPrizeMoney += rank.getPrizeMoney();
+        }
```
Stream을 사용해 볼까요? :)

tco0427 replied:
넵!!

```java
int totalPrizeMoney = ranks.keySet()
                .stream()
                .mapToInt(Rank::getPrizeMoney)
                .sum();
```

보다 가독성 있는 코드가 되는 것 같습니다!!

wbluke commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,49 @@
+        System.out.println("4개 일치 (50000원)-" + ranks.getOrDefault(FORTH, 0) + "개");
+        System.out.println("5개 일치 (1500000원)-" + ranks.getOrDefault(THIRD, 0) + "개");
+        System.out.println("5개 일치, 보너스 볼 일치(30000000원)-" + ranks.getOrDefault(SECOND, 0) + "개");
+        System.out.println("6개 일치 (2000000000원)-" + ranks.getOrDefault(FIRST, 0) + "개");
```
상금에 대한 부분도 Rank를 활용해야 하지 않을까요? Rank별 상금이 변경된다면? :)

tco0427 replied:
오..언급해주신 변경사항은 고려해보지 못했네요..앞서 피드백 주신 `사용하지 않는 필드`인 
```java
private final int prizeMoney;
```
위 필드를 사용하여 변경사항이 view에는 영향이 가지 않도록 수정해보았습니다!

wbluke commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,49 @@
+        System.out.println("5개 일치 (1500000원)-" + ranks.getOrDefault(THIRD, 0) + "개");
+        System.out.println("5개 일치, 보너스 볼 일치(30000000원)-" + ranks.getOrDefault(SECOND, 0) + "개");
+        System.out.println("6개 일치 (2000000000원)-" + ranks.getOrDefault(FIRST, 0) + "개");
+        System.out.printf("총 수익률은 %.2f 입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)", yield);
```
`(기준이 1이기 때문에 결과적으로 손해라는 의미임)`
이 멘트는 요구사항 설명을 위한 문구라 빼는 것이 좋겠어요.
사용자가 수익을 냈는데 위 문구를 보면 당황하지 않을까요? ㅎㅎ
클라이언트 사이드에서 고려하는 것도 굉장히 중요한 포인트 중 하나입니다 :)

tco0427 replied:
맞는 말씀이라고 생각합니다..!! 출력문구를 복붙하다보니..놓친 것 같습니다..😅

wbluke commented on src/test/java/lotto/domain/LottoTicketsTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings("NonAsciiCharacters")
```
테스트 클래스를 만들 때마다 달아줘야 할 것 같은데요. 둘 중 하나로 고민해 보셔요 :)
- DisplayName이 있으니 메서드 이름은 영어로 짓는다.
- 인텔리제이 설정으로 해결 가능

tco0427 replied:
맞습니다...저도 빠트리는 Class들이 빈번히 존재했습니다..저는 다른 사람과 같이 코딩한다고 생각했을 때 IDE 설정이 안된 동료도 있을 수 잇으므로 영어로 된 메소드명을 가지는 것이 좋을 것 같습니다..!! 수정해보았습니다.

wbluke reviewed (changes requested):
안녕하세요 디우~! 로또 미션 리뷰를 맡은 닉입니다 🙌 
일급컬렉션 사용해서 잘 구현해 주셨네요 ㅎㅎ 질문도 좋은 질문이네요.
질문에 대한 답변 포함해서 코멘트 남겼으니 확인 후 수정 부탁드려요 🙂

tco0427 committed:
refactor: Money -> LottoPurchaseMoney

좀 더 명확하고 직관적인 `LottoPurchaseMoney` 라는 네이밍으로 수정하였습니다. 이를 통해 cline가 해당 도메인에 대한 정보가 없어도 어떤 객체인지 명확하게 알 수 있습니다.

tco0427 committed:
refactor: `LottoPurchaseMoney` eq & hc 재정의

LottoPurchaseMoney는 특정 값을 나타내는 VO 이므로 equals & hash code를 재정의해주었습니다.

tco0427 committed:
feat: 로또 숫자 관리를 위한 VO 인 LottoNumber 정의

로또 숫자 관리를 위한 LottoNumber 라는 VO를 정의하였습니다.

tco0427 committed:
refactor: VO 패키지로 분리하여 관리

VO에 해당하는 LottoPurchaseMoney 와 LottoNumber 를 위한 VO 패키지를 domain 하위에 두어 관리하도록 하였습니다.

tco0427 committed:
refactor: 로또 구입금액 변환에 대한 책임을 InputView로 전환

로또 구입 금액을 입력받으면 String 형태인데 이를 정수(int)로 변환하는 책임을 view로 이동시켰습니다.

tco0427 committed:
fix: LottoPurchaseMoney 검증 추가

LottoPurchaseMoney 생성시 빠진 validateCriterion 부분을 추가하였습니다.

tco0427 committed:
feat: LottoNumber 리스트의 셔플을 위해 Comparable 구현

LottoNumber 의 정렬을 위해 Comparable 인터페이스를 구현하도록 수정하였습니다.

tco0427 committed:
fix: LottoNumberGenerator 에 LottoNumber 적용에 따른 수정 진행

LottoNumberGenerator 에서 List<Integer> 가 아닌 List<LottoNumber> 를 반환하도록 수정함에 따라 변경되는 사항을 함께 수정하였습니다.

tco0427 committed:
fix: Rank `NOTHING` 오타 수정

NOTHING 오타를 수정하였습니다.

tco0427 committed:
refactor: WinningNumbers `getCorrectCount` 리팩터링

int형으로 타입캐스팅으로 통한 불필요한 리스트 생성을 제거하였습니다.

tco0427 committed:
refactor: 매직 넘버에 대해 상수 적용

매직 넘버를 사용함으로 의미가 불분명해지는 부분에 대해서 상수를 적용하였습니다.

tco0427 committed:
refactor: LottoNumber 사용에 따른 수정

OutputView 에서 LottoNumber를 사용하므로 수정하였습니다.

tco0427 committed:
refactor: OutputView 상수 네이밍 변경

LOTTO_ENDFIX -> LOTTO_SUFFIX 로 변경하였습니다.

tco0427 committed:
refactor: 정수 입력에 대한 테스트 InputView로 전환

정수 입력 검증에 대한 책임을 view로 옮김으로써 정수가 아닌 입력에 대한 테스트를 WinningNumbersTest에서 할 수 없으므로 삭제하였습니다.

tco0427 committed:
refactor: LottoNumber 사용에 따른 WinningNumber 수정

LottoNumber 수정에 따라 WinningNumber 를 수정하게 되었고, 그에 따른 테스트 코드 및 view 를 함께 수정하였습니다.

tco0427 committed:
refactor: 당첨 번호 입력에 유연함을 가지도록 view 수정

당첨 번호 입력 시 ", " 가 아닌 "," 로 구분하더라도 정상 입력이 될 수 있도록 수정하였습니다.

tco0427 committed:
refactor: 로또 티켓 생성시 검증하는 기능을 추가

로또 티켓 생성시 다음을 검증하도록 수정해보았습니다.
* 6개의 숫자를 생성하는지
* 중복되는 숫자가 없는지

tco0427 committed:
refactor: 상수와 필드 구분

상수와 필드 사이 공백을 통해 구분해줄 수 있도록 하였습니다.

tco0427 committed:
refactor: LottoResult `calculateYield` 스트림 사용

Stream을 이용하여 보다 가독성 있는 코드로 리팩터링해보았습니다.

tco0427 committed:
refactor: OutputView 리팩터링

`(기준이 1이기 대문에 결과적으로 손해라는 의미임` 이라는 문구는 요구사항을 위한 메시지이고 사용자에게 보여주기 위한 메시지가 아니므로 삭제하였습니다.
또한 상금에 대한 부분도 Rank의 getPrizeMoney() 를 활용하도록 하여 상금 변화에 대한 View 변화를 최소화하도록 하였습니다.

tco0427 committed:
refactor: LottoTicket 매직넘버에 상수 적용

`로또 티켓에서 숫자는 6개여야 한다`를 검증하는 과정에서 사용하는 매직넘버에 대해서 상수 처리 해주었습니다.

tco0427 committed:
refactor: 상수와 필드 구분

공백을 통해서 상수와 필드를 구분해주었습니다.
또한 접근 제어자도 private 으로 변경해주었습니다.

tco0427 committed:
refactor: generator 네이밍 변경

기존 LottoNumberGenerator 는 `로또 번호 생성기` 라는 의미가 강하므로 6자리의 로또 티켓 하나를 생성해준다는 의미에서 LottoTicketGenerator 로 네이밍을 변경하였습니다.

tco0427 committed:
refactor: Rank 상수 private

Rank 상수에 대해서 private 접근제어자로 제한하고 기존 필드와 구분을 위해 공백을 추가해주었습니다.

tco0427 committed:
refactor: WinningNumbers 상수 private

tco0427 committed:
refactor: OutputView 상수 private

tco0427 committed:
refactor: 불필요한 import 구문 제거

tco0427 committed:
refactor: LottoTicket 테스트 코드 추가

6개 검증 및 중복 검증을 위한 단위테스트 코드 구현

tco0427 committed:
refactor: @SuppressWarnings 대신 영어 이름의 메소드명 사용

@SuppressWarnings을 사용하면 모든 테스트 클래스마다 추가해줘야한다는 단점이 존재한다. 테스트의 의도를 DisplayName을 통해서 충분히 확인할 수 있으므로 영어로 된 메소드명을 사용하기로 하였다.

wbluke commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -1,37 +1,63 @@
         String inputWinningNumbers = scanner.nextLine();
 
-        return Arrays.asList(inputWinningNumbers.split(SEPARATOR));
+        List<String> strings = Arrays.asList(inputWinningNumbers.replace(" ", "").split(SEPARATOR));
```
이것도 괜찮지만 replace()는 읽는 사람이 해석을 해야한다는 점에서, split() 후에 trim()하는 것도 괜찮을 것 같아요 :)

tco0427 replied:
동의합니다..!! 개선해보도록 하겠습니다!

wbluke commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -12,13 +13,14 @@ public LottoResult(Map<Rank, Integer> ranks) {
+    public double calculateYield(LottoPurchaseMoney lottoPurchaseMoney) {
+        int price = lottoPurchaseMoney.getPrice();
+
+        int totalPrizeMoney = ranks.keySet()
+                .stream()
+                .mapToInt(Rank::getPrizeMoney)
+                .sum();
```
위와 같이 길어지는 경우 메서드로 추출해 추상화를 하면 좋을 것 같아요 :)

tco0427 replied:
추상화 정도를 비슷하게 맞추면 좋다는 이야기를 `클린코드`에서 읽어본 것 같습니다..추상화하니 다른 코드와 추상화 정도도 비슷해지고 가독성 있는 것 같습니다! 따라서 저도 닉 의견에 동의합니다!! 개선해보겠습니다!!

wbluke commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -39,11 +58,11 @@ public void printLottoResultMessage() {
-        System.out.println("5개 일치, 보너스 볼 일치(30000000원)-" + ranks.getOrDefault(SECOND, 0) + "개");
-        System.out.println("6개 일치 (2000000000원)-" + ranks.getOrDefault(FIRST, 0) + "개");
-        System.out.printf("총 수익률은 %.2f 입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)", yield);
+        System.out.println("3개 일치 (" + FIFTH.getPrizeMoney() + "원)-" + ranks.getOrDefault(FIFTH, 0) + "개");
+        System.out.println("4개 일치 (" + FORTH.getPrizeMoney() + "원)-" + ranks.getOrDefault(FORTH, 0) + "개");
+        System.out.println("5개 일치 (" + THIRD.getPrizeMoney() + "원)-" + ranks.getOrDefault(THIRD, 0) + "개");
+        System.out.println("5개 일치, 보너스 볼 일치(" + SECOND.getPrizeMoney() + "원)-" + ranks.getOrDefault(SECOND, 0) + "개");
+        System.out.println("6개 일치 (" + FIRST.getPrizeMoney() + "원)-" + ranks.getOrDefault(FIRST, 0) + "개");
```
음 String.format()을 사용해보면 어떨까요? 메시지 형태가 한눈에 들어오게 할 수 있을 것 같아요 :)

tco0427 replied:
동의합니다..!! 개선해보니 조금 더 가독성 있는 출력구문이 된 것 같습니다!

wbluke commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -19,16 +24,30 @@ public void printTotalCount(int totalCount) {
+    }
+
+    private String getLottoTicketInfo(String ticketsInfo, List<LottoNumber> lottoNumbers) {
+        ticketsInfo = ticketsInfo + LOTTO_PREFIX;
+
+        for (LottoNumber lottoNumber : lottoNumbers) {
+            ticketsInfo = ticketsInfo + lottoNumber.getNumber() + SEPARATOR;
         }
 
+        ticketsInfo = ticketsInfo.substring(0, ticketsInfo.length()-2);
+        ticketsInfo = ticketsInfo + LOTTO_SUFFIX;
+
```
문자열 합산이 빈번하게 일어나고 있는데요, StringBuilder를 사용해보면 어떨까요?
적용하시면서 StringBuffer와의 차이도 같이 정리해보시면 좋고요 :)

tco0427 replied:
StringBuilder를 적용하여 저장공간이나 속도 측면에서 이점을 가질 수 있도록 하였습니다.
StringBuffer와의 차이는 쓰레드 환경에서 안전한지 여부라고 생각합니다. StringBuffer의 경우에 동기화가 가능하여 thread-safe하지만 속도가 느리기 때문에 String -> StringBuffer로의 이점이 크게 없다고 생각하여 StringBuilder를 사용하였습니다.

wbluke commented on src/main/java/lotto/domain/vo/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+
+    private static final int MINIMUM_LOTTO_NUMBER = 1;
+    private static final int MAXIMUM_LOTTO_NUMBER = 45;
+    private static final List<LottoNumber> LOTTO_TOTAL_NUMBERS = IntStream.rangeClosed(MINIMUM_LOTTO_NUMBER, MAXIMUM_LOTTO_NUMBER)
+            .mapToObj(LottoNumber::new)
+            .collect(toList());
+
+    private final int number;
+
+    public LottoNumber(int number) {
+        validateNumberBoundary(number);
+        this.number = number;
+    }
```
이 부분에서 고민이 필요해 보여요.
일단 1 ~ 45의 로또 숫자를 미리 생성해서 캐싱해놓는다는 관점에서는 잘 하셨어요 👍 
다만 그렇다면 다음 사항들을 추가로 고려해서 리팩토링 해보면 좋을 듯 해요.
- LottoNumber를 생성하려는 경우, 이미 생성되어 캐싱된 인스턴스가 있을 경우 새로 만들지 않고 기존 인스턴스를 반환하도록 한다.
  - 혹은 LottoNumber 내부에서가 아니라, LottoNumber 생성을 전담하는 Pool을 별도로 만들어도 좋다.
- 캐싱하는 자료구조가 List인데, 더 나은 자료구조는 없을까?

wbluke reviewed (approved):
피드백 반영 잘 해주셨네요 👍 
추가 코멘트 남겼고, 기존 코멘트에도 고민할 부분이 아직 보여서 답글 달았으니 같이 확인해주세요!
2단계 진행하시면서 같이 리팩토링 부탁드려요 🙂
